src/order/hom/complete_lattice.lean

```diff
@@ -0,0 +1,449 @@
+/-
+Copyright (c) 2022 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+-/
+import order.complete_lattice
+import order.hom.lattice
+
+/-!
+# Complete lattice homomorphisms
+
+This file defines frame homorphisms and complete lattice homomorphisms.
+
+We use the `fun_like` design, so each type of morphisms has a companion typeclass which is meant to
+be satisfied by itself and all stricter types.
+
+## Types of morphisms
+
+* `Sup_hom`: Maps which preserve `⨆`.
+* `Inf_hom`: Maps which preserve `⨅`.
+* `frame_hom`: Frame homomorphisms. Maps which preserve `⨆` and `⊓`.
+* `complete_lattice_hom`: Complete lattice homomorphisms. Maps which preserve `⨆` and `⨅`.
+
+## Typeclasses
+
+* `Sup_hom_class`
+* `Inf_hom_class`
+* `frame_hom_class`
+* `complete_lattice_hom_class`
+-/
+
+open function
+
+variables {F α β γ δ : Type*} {ι : Sort*} {κ : ι → Sort*}
+
+/-- The type of `⨆`-preserving functions from `α` to `β`. -/
+structure Sup_hom (α β : Type*) [has_Sup α] [has_Sup β] :=
+(to_fun   : α → β)
+(map_Sup' (s : set α) : to_fun (Sup s) = Sup (to_fun '' s))
+
+/-- The type of `⨅`-preserving functions from `α` to `β`. -/
+structure Inf_hom (α β : Type*) [has_Inf α] [has_Inf β] :=
+(to_fun   : α → β)
+(map_Inf' (s : set α) : to_fun (Inf s) = Inf (to_fun '' s))
+
+/-- The type of frame homomorphisms from `α` to `β`. They preserve `⊓` and `⨆`. -/
+structure frame_hom (α β : Type*) [complete_lattice α] [complete_lattice β] extends Sup_hom α β :=
+(map_inf' (a b : α) : to_fun (a ⊓ b) = to_fun a ⊓ to_fun b)
+
+/-- The type of complete lattice homomorphisms from `α` to `β`. -/
+structure complete_lattice_hom (α β : Type*) [complete_lattice α] [complete_lattice β]
+  extends Sup_hom α β :=
+(map_Inf' (s : set α) : to_fun (Inf s) = Inf (to_fun '' s))
```

Comment:
Note to self: why do we need these since `Inf` can be defined in terms of `Sup`? We don't know that `f '' {x | x ≤ s} = {y | y ≤ f '' s}`!

Reply:
Because that's wrong, right? Preserving limits is not the same as preserving colimits.

Reply:
Indeed! I was answering my own question in the previous comment, a fact I have to re-convince myself of with alarming regularity..